pkg/registries/registries.go

```diff
@@ -71,55 +76,127 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
 			Source:  source,
-			Mirrors: sortedRepos,
-		})
+			Mirrors: mirrors,
+		}
+		if mirrorBlockSource[source] {
+			imageDigestMirror.MirrorSourcePolicy = apicfgv1.NeverContactSource
+		}
+		res = append(res, imageDigestMirror)
+
 	}
 	return res, nil
 }
 
+func topoSortRepos(source string, ds *[]*[]apicfgv1.ImageMirror) ([]string, error) {
+	topoGraph := newTopoGraph()
+	for _, set := range *ds {
+		mirrors := *set
+		for i := 0; i+1 < len(mirrors); i++ {
+			topoGraph.AddEdge(string(mirrors[i]), string(mirrors[i+1]))
+		}
+		sourceInGraph := false
+		for _, m := range mirrors {
+			if string(m) == source {
+				sourceInGraph = true
+				break
+			}
+		}
+		if !sourceInGraph {
+			// The build of mirrorSets guarantees len(set.Mirrors) > 0.
+			topoGraph.AddEdge(string(mirrors[len(mirrors)-1]), source)
+		}
+		// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (every mirror set contains at least one non-source mirror,
+		// so there are no unconnected nodes that we would have to add separately from the edges).
+	}
+	sortedRepos, err := topoGraph.Sorted()
+	if err != nil {
+		return nil, err
+	}
+	if sortedRepos[len(sortedRepos)-1] == source {
+		// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
+		sortedRepos = sortedRepos[:len(sortedRepos)-1]
+	}
+	return sortedRepos, nil
+}
+
 // EditRegistriesConfig edits, IN PLACE, the /etc/containers/registries.conf configuration provided in config, to:
 // - Mark scope entries in insecureScopes as insecure (TLS is not required, and TLS certificate verification is not required when TLS is used)
 // - Mark scope entries in blockedScopes as blocked (any attempts to access them fail)
-// - Implement ImageContentSourcePolicy rules in icspRules.
+// - Implement ImageDigestMirrorSet rules in idmsRules.
+// - Implement ImageTagMirrorSet rules in itmsRules.
 // "scopes" can be any of whole registries, which means that the configuration applies to everything on that registry, including any possible separately-configured
 // namespaces/repositories within that registry.
 // or can be wildcard entries, which means that we accept wildcards in the form of *.example.registry.com for insecure and blocked registries only. We do not
 // accept them for mirror configuration.
 // A valid scope is in the form of registry/namespace...[/repo] (can also refer to sysregistriesv2.Registry.Prefix)
 // NOTE: Validation of wildcard entries is done before EditRegistriesConfig is called in the MCO code.
-func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScopes, blockedScopes []string, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) error {
+func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScopes, blockedScopes []string, idmsRules []*apicfgv1.ImageDigestMirrorSet, itmsRules []*apicfgv1.ImageTagMirrorSet) error {
```

Comment:
How is v1alpha1.ICSP going to be handled?

Considering there are so many users that are going to need updating to support he new CRs, I’d expect the ICSP handling to exist in the shared code (i.e. in this package), definitely not copy&pasted in every caller.